### ISSUE TITLE
SqlServer Query: double.NaN (or float) constant produces invalid SQL

### ISSUE THREAD

**Issue by bricelam** [dotnet/efcore#12998]
> SqlException : Invalid column name 'NaN'

There is no way to represent `NaN`, ~~`PositiveInfinity`, or `NegativeInfinity`~~ in Transact-SQL.

**Comment by ajcvickers** [comment]
Triage:
* Do what we need to do for spatial (likely based on feedback)
* Consider client eval or throw for 3.0.

**Comment by smitpatel** [comment]
Dependent on #13192

**Comment by GSPP** [comment]
Maybe send `0.0/0.0` to the server as a representation for `NaN`.

**Comment by bricelam** [comment]
Silly me, you'd never write `x == double.NaN` because it's always false. You'd have to write `double.IsNaN(x)`.

For spatial, it could be convenient if we translated `double.IsNaN(x.Geometry.Z)` to `[x].[Geometry].HasZ()`, but again I think it's low priority since if you're using Z or M, you likely always specify a Z or M value.

**Comment by ajcvickers** [comment]
There is no NaN in T-SQL. Should consider using null, client eval, or just not supported.

**Comment by ajcvickers** [comment]
Notes from triage: The literal generator for the provider/type/value combination should throw if asked to generate a literal for values that cannot be represented in the store. Se also #12797.

**Comment by bricelam** [comment]
This behavior is consistent with SqliteParameter which [throws](https://github.com/aspnet/EntityFrameworkCore/blob/68127d93b6d92775f5b006edf620ce8a6fa6d773/src/Microsoft.Data.Sqlite.Core/SqliteValueBinder.cs#L32-L35) if the value is `double.NaN`.
